crates/rustc_codegen_spirv/src/codegen_cx/entry.rs

```diff
@@ -264,25 +269,67 @@ impl<'tcx> CodegenCx<'tcx> {
                     ),
+                )
+            }
+
+            match inferred_storage_class_from_ty {
+                Some(inferred) if storage_class == inferred => self.tcx.sess.span_warn(
+                    storage_class_attr.span,
+                    "redundant storage class specifier, storage class is inferred from type",
                 ),
+                Some(inferred) => self
+                    .tcx
+                    .sess
+                    .struct_span_err(
+                        hir_param.span,
+                        &format!(
+                            "storage class {:?} was inferred from type, but {:?} was specified in attribute",
+                            inferred, storage_class
+                        ),
+                    )
+                    .span_note(
+                        storage_class_attr.span,
+                        &format!("remove storage class attribute to use {:?} as storage class", inferred),
+                    )
+                    .emit(),
+                None => (),
             }
-        } else {
-            match *layout.ty.kind() {
-                TyKind::Ref(_, pointee_ty, hir::Mutability::Mut) => {
-                    (pointee_ty, StorageClass::Output)
-                }
 
-                TyKind::Ref(_, pointee_ty, hir::Mutability::Not) => self.tcx.sess.span_fatal(
+            storage_class
+        });
+        // If storage class was not inferred nor specified, compute the default (i.e. input/output)
+        let storage_class = inferred_storage_class_from_ty
+            .or(attr_storage_class)
+            .unwrap_or_else(|| match (is_ref, mutbl) {
+                (false, _) => StorageClass::Input,
+                (true, hir::Mutability::Mut) => StorageClass::Output,
+                (true, hir::Mutability::Not) => self.tcx.sess.span_fatal(
                     hir_param.span,
                     &format!(
                         "invalid entry param type `{}` (expected `{}` or `&mut {1}`)",
-                        layout.ty, pointee_ty
+                        layout.ty, value_ty
                     ),
                 ),
+            });
 
-                _ => (layout.ty, StorageClass::Input),
-            }
-        };
+        (spirv_ty, storage_class)
+    }
+
+    fn declare_parameter(
+        &self,
+        layout: TyAndLayout<'tcx>,
+        hir_param: &hir::Param<'tcx>,
+        decoration_locations: &mut HashMap<StorageClass, u32>,
+    ) -> (Word, StorageClass) {
+        let attrs = AggregatedSpirvAttributes::parse(self, self.tcx.hir().attrs(hir_param.hir_id));
+
+        // FIXME(eddyb) attribute validation should be done ahead of time.
+        // FIXME(eddyb) also take into account `&T` interior mutability,
+        // i.e. it's only immutable if `T: Freeze`, which we should check.
+        // FIXME(eddyb) also check the type for compatibility with being
+        // part of the interface, including potentially `Sync`ness etc.
```

Comment:
These comments belong in `infer_param_ty_and_storage_class` (they pertain to the `TyKind::Ref` part of the Rust type, more or less).